apis/apiextensions/v1/composition_types.go

```diff
@@ -213,6 +222,16 @@ type Patch struct {
 	// +optional
 	FromFieldPath *string `json:"fromFieldPath,omitempty"`
 
+	// FromFieldPathStrategy determines how to patch from a field path. The
```

Comment:
Is there a specific reason this isn't named `Strategy`? I can see this strategy being applied for all operations of the patch, i.e. not only fetching the value from field path but also not being able to write the value to the destination.

Reply:
I think there's value in distinguishing the source and destination here. For example I could imagine "to field path" strategies like "Replace" vs "Merge". I considered trying to model that all under a `strategy` key too, like:

```yaml
- type: FromCompositeFieldPath
  fromFieldPath: metadata.annotations
  toFieldPath: metadata.annotations
  strategy:
    fromFieldPath: Required
    toFieldPath: Merge
```

I figured in this case that would look instead like:

```yaml
- type: FromCompositeFieldPath
  fromFieldPath: metadata.annotations
  fromFieldPathStrategy: Required
  toFieldPath: metadata.annotations
  toFieldPathStrategy: Merge
```

I don't feel strongly which one we go with though.

Reply:
I agree that strategy could be different between the source and the destination, so separating them makes sense. 

I find the first example a little easier to read as the fields look very similar to the existing fields in the second example.

Is there a possibility that future patch types might not require either the `fromFieldPath` or `toFieldPath` settings but still want to define a strategy for whatever their source or destination is?

If so, maybe the field names could be more generic (e.g. `sourceStrategy` or `fromStrategy`).

The only example I could think of there would be a patch type that retrieves something from entirely outside the k8s API, or something that reads an entire resource (so does not use `fromFieldPath`), but you might still want to define a strategy based on the existence of that source. I don't know how likely that is but thought it worth mentioning just in case 😃

Reply:
It feels like `Optional|Required` choices are not in the same class with `Merge|Replace`. Like I might want a patch `toFieldPath` to be `Required` as well as `Merge`. Also, I'm just realizing that `patch strategy` is actually used in other contexts, like patch strategies in kubectl. So, it feels like a different name and field for `Optional|Required` choice could make sense. The `strategy` would be preserved for the actual strategy of the patch, i.e. _how_ the patch should be done instead of whether to enforce it or not.

Reply:
> I find the first example a little easier to read as the fields look very similar to the existing fields in the second example.

I'm happy to take that approach instead.

> Is there a possibility that future patch types might not require either the fromFieldPath or toFieldPath settings but still want to define a strategy for whatever their source or destination is?

There is, but I figured we could add strategies as appropriate, rather than trying to come up with a generic concept before we know what exactly it's for. I prefer a "from [field path] strategy" over a "source strategy" because the similar naming makes the relationship between the two fields obvious. I don't think it would be so bad that if in future we added a `type: FromWidget` patch we add a `strategy.fromWidget` next to the `strategy.fromFieldPath`.

> It feels like Optional|Required choices are not in the same class with Merge|Replace. Like I might want a patch toFieldPath to be Required as well as Merge.

"Strategy" is a pretty open ended word. To me it can fit both "what is our strategy for reading the value" (e.g. is it required) and "what is our strategy for writing the value" (e.g. should it be merged). The concepts "read" and "write" in our API are usually something like "from" and "to".

I'm going to update the API to the "nested" option above for now. I don't want to quash debate around the best API here, but I would love to get this merged before our feature freeze on Monday if possible. Let me know if anyone has any concerns with the proposed approach - ideally with a counterproposal as to how this might work. :)

Reply:
> It feels like Optional|Required choices are not in the same class with Merge|Replace. Like I might want a patch toFieldPath to be Required as well as Merge.

It just occurred to me that we might be interpreting this differently. Are you thinking that `FromFieldPathStrategy` would be the strategy used for all patches of `type: FromCompositeFieldPath`, while `ToFieldPathStrategy` would be the strategy used for all patches of `type: ToCompositeFieldPath`?

My intention was that these strategies would correspond more the `fromFieldPath:` and `toFieldPath:` fields, regardless of which patch type is used. If you look at it this way I think the current design satisfies your criteria, e.g.:

```yaml
# The 'from' path is required, and we merge into the 'to' path.
- type: FromCompositeFieldPath
  fromFieldPath: metadata.annotations
  toFieldPath: metadata.annotations
  strategy:
    fromFieldPath: Required
    toFieldPath: Merge
# Same as the above, but in the other direction.
- type: ToCompositeFieldPath
  fromFieldPath: metadata.annotations
  toFieldPath: metadata.annotations
  strategy:
    fromFieldPath: Required
    toFieldPath: Merge
```

Reply:
> It just occurred to me that we might be interpreting this differently. Are you thinking that FromFieldPathStrategy would be the strategy used for all patches of type: FromCompositeFieldPath, while ToFieldPathStrategy would be the strategy used for all patches of type: ToCompositeFieldPath?

Not really, I think we're on the same page.

My worry is that we might want `toFieldPath` to have both `Required` and `Merge` strategies so to say, because the possible `strategy` types we're expecting to exist even now (merge, required, optional, replace) are not exclusive; that's why having those types under one umbrella could limit us.

The following is an example of how we can make `strategy` specific to patching operation strategy (`setValue`) we'll follow and have more room for details about the chosen strategy. This specific example shows how to merge arrays and what to use as identifier to decide which element should merge to which. The field `required` would then capture the whole pipeline, including fetching, transform operations and setting the final value; if any of them fails, `required` dictates that we can't continue.
```yaml
- type: FromCompositeFieldPath
  fromFieldPath: spec.someArray
  toFieldPath: spec.some.otherArray
  required: true
  strategy:
    type: Merge
    merge:
      identifier: ".name"
```

Additionally, in this way we tie these properties to the common operations of all kinds of patches we do (get, transform, set) rather than fields (`fromFieldPath`, `toFieldPath` or even `fromMultipleFieldPaths`) which may change with the type.

Reply:
Let me try take a step back.

My aim with this PR is to allow Composition authors to mark the `fromFieldPath` of a patch as required. This is something that we're confident would be helpful to people, for example by ensuring we wait until an XR's `spec.claimRef.namespace` is set before we try to patch from it.

I didn't pick the name "strategy" with the API server concept of a "patch strategy" (e.g. merge patch) in mind, but I can see the confusion there. I picked it because it's a pretty generic way to say "how to do a thing", and [API conventions recommend](https://github.com/kubernetes/community/blob/6254e11696ac0886d554021a27c8179b459502c6/contributors/devel/sig-architecture/api-conventions.md#primitive-types) avoiding the obvious route of adding a `required: true` boolean:

> Think twice about bool fields. Many ideas start as boolean but eventually trend towards a small set of mutually exclusive options. Plan for future expansions by describing the policy options explicitly as a string type alias (e.g. TerminationMessagePolicy).

Now as we try to figure out how we should model whether or not the `fromFieldPath` is required we're factoring in some other functionality we _may_ want to add in future, including allowing Composition authors to control:

1. How the `toFieldPath` should be patched - e.g. merged or replaced.
1. Whether the `toFieldPath` should be required. (I'm guessing this means "only patch this field if it already exists"?)
1. Whether other patch-related-stuff like transforms are required.

I think we have varying levels of confidence that folks will want to control these things. Personally, 1 seems pretty likely to me but I can't immediately think of any reason someone would want to control 2 or 3. Given this, I'd like to find a way to let folks control whether the `fromFieldPath` is required that doesn't conflate the requiredness of the `fromFieldPath` with the requiredness of other parts of the patch, but that also leaves a path open to modelling everything we've touched on in this issue if and when we need to.

Perhaps we can move forward here by calling the 'strategy' field 'policy' or something equally generic that folks won't associate with API server patch strategies. So something like:

```yaml
- type: FromCompositeFieldPath
  fromFieldPath: spec.claimRef.namespace
  toFieldPath: spec.providerConfigRef.name
  policy:
    fromFieldPath: Required
```

This avoids conflating 'the from field path is required' with 'everything about this patch is required'. It also leaves open space to separately model everything else we've discussed, e.g.:

```yaml
- type: FromCompositeFieldPath
  fromFieldPath: spec.claimRef.namespace
  toFieldPath: spec.providerConfigRef.name
  policy:
    fromFieldPath: Required
    toFieldPath: Required
    transforms: Required
    someOtherThingWeHaventThoughtOfYet: Optional
  strategy:
    type: Merge
    merge:
      identifier: ".name"
```

Reply:
> I'd like to find a way to let folks control whether the fromFieldPath is required that doesn't conflate the requiredness of the fromFieldPath with the requiredness of other parts of the patch, but that also leaves a path open to modelling everything we've touched on in this issue if and when we need to.

The cases where transform interact with external systems or many inputs could be candidate for this. I can't really come up with an example for requiredness of `setValue`, i.e. `toFieldPath`, other than schema error like you try to set an array to a string field. So, my thought was that if I want `fromFieldPath` to be required, chances are pretty high that I want other operations to be required as well because requiredness for me is about having the final value in the managed resource and blocking if that doesn't happen, so transform/`setValue` failures should also block the rendering.

<img width="744" alt="image" src="https://user-images.githubusercontent.com/7584126/107928577-8d7e4f00-6f89-11eb-9890-9070888b5a2d.png">

This looks pretty good to me. My main worry was that strategies and requiredness are not mutually exclusive to be in the same field and this separation seems to solve it pretty well. I can't really think of much scenarios but more granularity about which operations are required could also come in handy.